swarm-derive/src/lib.rs

```diff
@@ -485,23 +506,27 @@ fn build_struct(ast: &DeriveInput, data_struct: &DataStruct) -> TokenStream {
             out_handler.unwrap_or(quote! {()}) // TODO: See test `empty`.
         };
 
-        let generate_event_match_arm = if event_process {
-            quote! {
-                std::task::Poll::Ready(#network_behaviour_action::GenerateEvent(event)) => {
-                    #net_behv_event_proc::inject_event(self, event)
-                }
-            }
-        } else {
+        let generate_event_match_arm = {
+            // If the `NetworkBehaviour`'s `OutEvent` is generated by the derive macro, wrap the sub
+            // `NetworkBehaviour` `OutEvent` in the variant of the generated `OutEvent`. If the
+            // `NetworkBehaviour`'s `OutEvent` is provided by the user, use the corresponding `From`
+            // implementation.
+            let into_out_event = if out_event_definition.is_some() {
+                quote! { #out_event_name::#event_variant(event) }
+            } else {
+                quote! { event.into() }
+            };
```

Comment:
Good idea!

Reply:
Initially I wanted to use `From` in both cases, though Rust has issues with the implementation below:

``` rust
impl From<<SomeBehaviour as NetworkBehaviour>::OutEvent> for GeneratedOutEvent
```

Namely with `<SomeBehaviour as NetworkBehaviour>::OutEvent` not being specific enough.